src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Text.Projection;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Venus
+{
+    internal sealed partial class ContainedDocument
+    {
+        public class DocumentServiceProvider : IDocumentServiceProvider
+        {
+            public static readonly IDocumentServiceProvider Instace = new DocumentServiceProvider();
+
+            private readonly SpanMapper _service;
+
+            private DocumentServiceProvider()
+            {
+                _service = new SpanMapper();
+            }
+
+            public TService GetService<TService>() where TService : IDocumentService
+            {
+                if (_service is TService service)
+                {
+                    return service;
+                }
+
+                return default;
+            }
+
+            private class SpanMapper : ISpanMapperService
+            {
+                public async Task<ImmutableArray<DocumentSpan>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken)
+                {
+                    var containedDocument = ContainedDocument.TryGetContainedDocument(document.Id);
+                    if (containedDocument == null)
+                    {
+                        return spans.Select(s => new DocumentSpan(document, s)).ToImmutableArray();
+                    }
+
+                    var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    var editorSnapshot = sourceText.FindCorrespondingEditorTextSnapshot() as IProjectionSnapshot;
+                    if (editorSnapshot == null)
+                    {
+                        return spans.Select(s => new DocumentSpan(document, s)).ToImmutableArray();
+                    }
+
+                    var builder = ArrayBuilder<DocumentSpan>.GetInstance();
+                    foreach (var span in spans)
+                    {
+                        foreach (var primarySpan in editorSnapshot.MapToSourceSnapshots(span.ToSpan()))
```

Comment:
Is this mapping in the correct direction? I would have presumed this was taking a location in the secondary buffer and mapped to the primary, and technically editorSnapshot doesn't have to be an editor snapshot at all. This might be relying on the fact that the data buffer/document buffer have the same contents but it's kinda strange...

Reply:
I believe it is on right direction.

..

this is old code we use (which I also verified it works)
TryMapSpanFromSecondaryBufferToPrimaryBuffer (http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Extensions/VsTextSpanExtensions.cs,12)

I basically ported editor implementation
http://index/?query=MapSecondaryToPrimarySpan&rightProject=Microsoft.VisualStudio.Editor.Implementation&file=VsTextBufferCoordinatorAdapter.cs&line=177

but, I might wrongly cast it to IProjectionBuffer. will take a look.

Reply:
I think name of editorSnapshot confusing it. I used editorSnapshot as a way to say it is not Roslyn Buffer. but I can rename it to secondaryBufferSnapshot.